addons/mail/static/src/web/messaging_menu/messaging_menu.js

```diff
@@ -117,39 +117,39 @@ export class MessagingMenu extends Component {
             if (a.correspondent === this.store.odoobot) {
                 return 1;
             }
-            if (!a.newestPersistentMessage?.datetime) {
+            if (a.hasNeedactionMessages && !b.hasNeedactionMessages) {
                 return -1;
             }
-            if (!b.newestPersistentMessage?.datetime) {
+            if (b.hasNeedactionMessages && !a.hasNeedactionMessages) {
+                return 1;
+            }
+            if (a.message_unread_counter > 0 && b.message_unread_counter === 0) {
+                return -1;
+            }
+            if (b.message_unread_counter > 0 && a.message_unread_counter === 0) {
+                return 1;
+            }
+            if (!a.newestPersistentMessage?.datetime) {
                 return 1;
             }
+            if (!b.newestPersistentMessage?.datetime) {
+                return -1;
+            }
             return b.newestPersistentMessage.datetime - a.newestPersistentMessage.datetime;
         });
         const previews = [];
         for (const thread of threads) {
             const { newestMessage, newestNeedactionMessage } = thread;
-            if (thread.is_pinned) {
-                const message = newestMessage;
+            if (thread.is_pinned || newestNeedactionMessage) {
                 previews.push({
                     id: `preview-${thread.localId}`,
-                    count: thread.message_unread_counter,
-                    imgUrl: thread.imgUrl,
-                    hasMarkAsReadButton: thread.message_unread_counter > 0,
-                    message,
-                    thread,
-                    isNeedaction: false,
-                });
-            }
-            if (newestNeedactionMessage) {
-                const message = newestNeedactionMessage;
-                previews.push({
-                    id: `preview-needaction-${thread.localId}`,
-                    count: thread.needactionMessages.length,
-                    imgUrl: message.module_icon,
-                    hasMarkAsReadButton: true,
-                    message,
+                    counter: thread.message_needaction_counter,
```

Comment:
why keep a counter key here if the template is giving another counter props?

Reply:
This whole formatter in JS could be moved to template now that there is only one occurrence?
The getter would just return the sorted/filtered threads?